cmd/nephe-controller/main.go

```diff
@@ -47,6 +56,124 @@ func init() {
+type secretValidator struct {
+	client  controllerclient.Client
+	decoder *admission.Decoder
+}
+
+// allowSecretUpdate returns nil only when secret name/namespace/key are same.
+func (v *secretValidator) allowSecretUpdate(new *corev1.Secret, old *corev1.Secret, key string) error {
+	setupLog.Info("Checking secrets are changed")
+	if new.Name != old.Name {
+		return fmt.Errorf("secret name cannot be changed, old: %s , new: %s", new.Name, old.Name)
+	}
+	if new.Namespace != old.Namespace {
+		return fmt.Errorf("secret namespace cannot be changed, old: %s , new: %s", new.Namespace, old.Namespace)
+	}
+	if changed := string(new.Data[key]) != string(old.Data[key]); changed {
+		return fmt.Errorf("secret data %s cannot be changed, old: %s , new: %s", key, new.Name, old.Name)
+	}
+	return nil
+}
+
+// fetchCPABySecret returns nil only when the secret is not used by cpa
+// otherwise, an appropriate error message and cpa object will be returned.
+func (v *secretValidator) fetchCPABySecret(newSecret *corev1.Secret, oldSecret *corev1.Secret, isDelete bool) (error, *crdv1alpha1.CloudProviderAccount) {
+	// TODO: Figure out how to get all CRs
+	cpaList := &crdv1alpha1.CloudProviderAccountList{}
+	options := &controllerclient.ListOptions{Limit: 1000}
+	err := v.client.List(context.TODO(), cpaList, options)
+	if err != nil {
+		return fmt.Errorf("failed to get cpa list, err:%v", err), nil
+	}
+
+	for _, cpa := range cpaList.Items {
+		setupLog.Info("Checking cpa", "name", cpa.Name, "namespace", cpa.Namespace)
+		if cpa.Spec.AWSConfig != nil {
+			if cpa.Spec.AWSConfig.SecretRef.Name == oldSecret.Name && cpa.Spec.AWSConfig.SecretRef.Namespace == oldSecret.Namespace {
+				if isDelete {
+					// Delete is not allowed.
+					setupLog.Info("Cannot delete secret", "name", oldSecret.Name, "cpa", oldSecret.Name)
+					return fmt.Errorf("cannot delete secret %s", oldSecret.Name), &cpa
+				}
+
+				if err = v.allowSecretUpdate(newSecret, oldSecret, cpa.Spec.AWSConfig.SecretRef.Key); err != nil {
+					return err, &cpa
+				}
+			}
+		}
+
+		if cpa.Spec.AzureConfig != nil {
+			if cpa.Spec.AzureConfig.SecretRef.Name == oldSecret.Name && cpa.Spec.AzureConfig.SecretRef.Namespace == oldSecret.Namespace {
+				if isDelete {
+					// Delete is not allowed.
+					setupLog.Info("Cannot delete secret", "name", oldSecret.Name, "cpa", oldSecret.Name)
+					return fmt.Errorf("cannot delete secret %s", oldSecret.Name), &cpa
+				}
+				if err = v.allowSecretUpdate(newSecret, oldSecret, cpa.Spec.AzureConfig.SecretRef.Key); err != nil {
+					return err, &cpa
+				}
+			}
+		}
+	}
+	return nil, nil
+}
+
+// Handle handles admission requests for secret.
+func (v *secretValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	setupLog.Info("Received webhook", "req", req, "operation", req.Operation)
+	switch req.Operation {
+	case admissionv1.Create:
+		return admission.Allowed("")
+	case admissionv1.Update:
+		newSecret := &corev1.Secret{}
```

Comment:
Create separate routine for update and delete case..

Reply:
Done